py-rse/style.Rmd

```diff
@@ -287,755 +349,875 @@ n = count(lines)
-readers will assume that it might change each time the code is executed.
+### Replace code with data.
 
-Our second refactoring,
-"hoist repeated calculation out of loop",
-tells us to move the repeated calculation out of the loop or function.
-Doing this signals that its value is always the same.
-And by naming that common value,
-you help readers understand what its purpose is.
+It is easier to understand and maintain lookup tables than complicated conditionals,
+so the "create lookup table" refactoring tells us to turn the latter into the former:
 
 ```python
-# ...before...
-for sample in signals:
-    output.append(2 * pi * sample / weight)
+# ...before..
+def count_vowels_and_consonants(text):
+    num_vowels = 0
+    num_consonants = 0
+    for char in text:
+        if char in VOWELS:
+            num_vowels += 1
+        elif char in CONSONANTS:
+            num_consonants += 1
+    return num_vowels, num_consonants
 ```
 
 ```python
 # ...after...
-scaling = 2 * pi / weight
-for sample in signals:
-    output.append(sample * scaling)
-```
+IS_VOWEL = {'a' : 1, 'b' : 0, 'c' : 0, ... }
+IS_CONSONANT = {'a' : 0, 'b' : 1, 'c' : 1, ... }
 
-### Replace tests with flags to clarify repeated tests
+def count_vowels_and_consonants(text):
+    num_vowels = num_consonants = 0
+    for char in text:
+        num_vowels += IS_VOWEL[char]
+        num_consonants += IS_CONSONANT[char]
+    return num_vowels, num_consonants
+```
 
-Novice programmers frequently write conditional tests like this:
+The more distinct cases there are,
+the greater the advantage lookup tables have over multi-branch conditionals.
+Those advantages multiply when items can belong to more than one category,
+in which case the table is often best written as a dictionary with items as keys
+and sets of categories as values:
 
 ```python
-if (a > b) == True:
-    # ...do something...
+LETTERS = {
+    'A' : {'vowel', 'upper_case'},
+    'B' : {'consonant', 'upper_case'},
+    # ...other upper-case letters...
+    'a' : {'vowel', 'lower_case'},
+    'b' : {'consonant', 'lower_case'},
+    # ...other lower-case letters...
+    '+' : {'punctuation'},
+    '@' : {'punctuation'},
+    # ...other punctuation...
+}
+
+def count_vowels_and_consonants(text):
+    num_vowels = num_consonants = 0
+    for char in text:
+        num_vowels += int('vowel' in LETTERS[char])
+        num_consonants += int('consonant' in LETTERS[char])
+    return num_vowels, num_consonants
 ```
 
-The comparison to `True` is unnecessary because `a > b` is a Boolean value
-that is itself either `True` or `False`.
-Like any other value,
-Booleans can be assigned to variables,
-and those variables can then be used directly in tests:
+The expressions used to update `num_vowels` and `num_consonants` make use of the fact that
+`in` produces either `True` or `False`,
+which the function `int` converts to either 1 or 0.
+We will explore ways of making this code more readable in the exercises.
```

Comment:
Is it worth mentioning somewhere that Python does not have a "case" operator, but that lookup tables or if/else are used for that instead.

Reply:
I think most of our readers will never have run into `case`.